fog/ingest/server/src/controller.rs

```diff
@@ -166,10 +213,12 @@ where
             result
                 .restore_state_from_summary(summary)
                 .unwrap_or_else(|err| {
-                    panic!(
-                        "Could not restore state from state file ({:?}), {:?}: {}",
+                    // Note: We don't panic or nuke the statefile because this server may have recovered
```

Comment:
This seems a bit fishy to me, because `restore_state_from_summary` can fail halfway through, and depending on whats inside it you might be left with a partially initialized object. It doesn't seem to be an issue right now, since I think the only thing that can modifies state and can fail is `set_peers_inner` but if more stuff is added this can result in subtle bugs.

Reply:
So I see your sentiment here -- I think part of the issue is that I'm reaching the conclusion that sometimes "partially initializing" the state is the right thing here.

For instance in the following scenario:
* All ingests went down
* This is the only one that still has the right key in its sealed key file
* When it restarts, restore_state fails, because during the call to confirm_backup, some of the peers are unreachable due to a network issue. This node therefore can't confirm that it wouldn't be a second active node, so it backs off.
* So restore_state_from_summary returns a peer checkup error

At this point, if we start in an idle state, that's fine, and something can later make us active if it figures out that we have the right key. Or something else could become active and we could follow it.

If we nuke the statefile and panic, then we lost the key.

If we panic without nuking the statefile, that would be fine in this scenario since we would try the peer checkup again and perhaps succeed, or determine that someone else has become the leader, and we should be a follower now.

Maybe the right answer is that we need to inspect the error and perform retries if it appears to be network related.

Reply:
I agree with you about the desirable effect. I am just not certain that the way it's currently implemented is optimal in getting us there. Inspecting the error might be a good path forward - perhaps even without retries, just if the error is network-related then go to `Idle` instead of crashing.

Reply:
I agree that this does suggest some sort of third state or separate flag since it is otherwise misleading if it isn't fully initialized such that we can actually make this idle instance the active one. Could we just set a flag indicating that this state is not ready for use as well as logging the error when it happens?